### ISSUE TITLE
Generic runtime representation 

### ISSUE THREAD

**Issue by tim2CF** [awakesecurity/gRPC-haskell#106]
Hello! Is there some sort of generic runtime representation of protobuf types generated from protobuf files? Something similar to this (but for protobuf) https://hackage.haskell.org/package/aeson-1.5.3.0/docs/Data-Aeson.html#t:Value

**Comment by Gabriel439** [comment]
@tim2CF: There is not a generic representation, as far as I know.

**Comment by tim2CF** [comment]
I meant not exactly GHC.Generics, but some sort of protobuf AST which can be used to fold protobuf messages in generic way. Or at least some sort of protobuf reflection API which can give information about message field types and indexes in terms of Haskell.

**Comment by tim2CF** [comment]
Well, maybe not data itself, but schema of data. Information about, types, names and indexes of messages in form of Haskell terms. Protobuf reflection API exists in some implementations. I'm not sure is it part of standard itself, but you for example take a look to other Haskell implementation here

https://hackage.haskell.org/package/proto-lens-0.7.0.0/docs/Data-ProtoLens-Message.html#g:1

It's quite useful in some cases, for example for generalized Arbitrary instance for any protobuf message

http://hackage.haskell.org/package/proto-lens-arbitrary-0.1.2.9/docs/src/Data.ProtoLens.Arbitrary.html#arbitraryMessage

**Comment by Gabriel439** [comment]
@tim2CF: Oh, so if all you need is the AST of the `.proto` file then our `proto3-suite` package supports that.  This module is the entry point to parsing the AST from a `.proto` file: https://hackage.haskell.org/package/proto3-suite-0.4.0.0/docs/Proto3-Suite-DotProto-Generate.html

**Comment by Gabriel439** [comment]
@tim2CF: Yeah, I believe the `dotProto` method will allow you to retrieve metadata about fields
